data/base/controller.php

```diff
@@ -147,50 +163,56 @@ public function register_processors() {
 	}
 
 	/**
-	 * Register internal endpoints.
+	 * Register index endpoint.
 	 */
-	protected function register_internal_endpoints() {
-		register_rest_route( $this->get_namespace(), '/' . $this->get_rest_base(), [
-			[
-				'methods' => WP_REST_Server::READABLE,
-				'callback' => array( $this, 'get_items' ),
-				'args' => [],
-				'permission_callback' => function ( $request ) {
-					return $this->get_permission_callback( $request );
-				},
-			],
-		] );
+	protected function register_index_endpoint() {
+		$this->register_endpoint( Endpoint\Index::class );
 	}
 
 	/**
 	 * Register endpoint.
 	 *
 	 * @param string $endpoint_class
 	 *
-	 * @return \Elementor\Data\Base\Endpoint
+	 * @return \Elementor\Data\Base\Endpoint|false
 	 */
 	protected function register_endpoint( $endpoint_class ) {
 		$endpoint_instance = new $endpoint_class( $this );
 
-		// TODO: Validate instance like in register_sub_endpoint().
+		if ( ! ( $endpoint_instance instanceof Endpoint ) ) {
+			trigger_error( 'Invalid endpoint instance.' );
+			return false;
+		}
 
-		$endpoint_route = $this->get_name() . '/' . $endpoint_instance->get_name();
+		$command = $endpoint_instance->get_full_command();
 
-		$this->endpoints[ $endpoint_route ] = $endpoint_instance;
+		if ( $endpoint_instance instanceof Endpoint\Index ) {
+			$this->index_endpoint = $endpoint_instance;
+		} else {
+			$this->endpoints[ $command ] = $endpoint_instance;
+		}
+
+		$this->register_endpoint_format( $command, $endpoint_instance );
+
+		return $endpoint_instance;
+	}
 
-		$command = $endpoint_route;
+	/**
+	 * Register endpoint format of controllers.
+	 *
+	 * @param string $command
+	 * @param \Elementor\Data\Base\Endpoint $endpoint_instance
+	 */
+	protected function register_endpoint_format( $command, $endpoint_instance ) {
 		$format = $endpoint_instance::get_format();
 
-		if ( $command ) {
-			$format = $command . '/' . $format;
-		} else {
-			$format = $format . $command;
+		// Means, the format includes the full-path.
+		if ( ! strstr( $format, '/{' ) ) {
```

Comment:
Replace all places to a full format & remove the if.

Reply:
Handled: https://github.com/elementor/elementor/pull/12820/commits/67cf8416bb7681a573ac13904e544eda3c4bf564